source/loader/ur_lib.cpp

```diff
@@ -206,4 +212,579 @@ urLoaderConfigSetCodeLocationCallback(ur_loader_config_handle_t hLoaderConfig,
+                return ""; // no ODS string matches this
+                break;
+            default:
+                return ""; // no ODS string matches this
+                break;
+            }
+        }();
+
+    // The std::map is sorted by its key, so this method of parsing the ODS env var
+    // alters the ordering of the terms, which makes it impossible to check whether
+    // all discard terms appear after all accept terms and to preserve the ordering
+    // of backends as specified in the ODS string.
+    // However, for single-platform requests, we are only interested in exactly one
+    // backend, and we know that discard filter terms always override accept filter
+    // terms, so the ordering of terms can be safely ignored -- in the special case
+    // where the whole ODS string contains at most one accept term, and at most one
+    // discard term, for that backend.
+    // (If we wished to preserve the ordering of terms, we could replace `std::map`
+    // with `std::queue<std::pair<key_type_t, value_type_t>>` or something similar.)
+    auto &maybeEnvVarMap = getenv_to_map("ONEAPI_DEVICE_SELECTOR", true);
+
+    // if the ODS env var is not set at all, then pretend it was set to the default
+    using EnvVarMap = std::map<std::string, std::vector<std::string>>;
+    EnvVarMap mapODS = maybeEnvVarMap.has_value() ? maybeEnvVarMap.value()
+                                                  : EnvVarMap{{"*", {"*"}}};
+
+    // the full BNF grammar can be found here:
+    // https://github.com/intel/llvm/blob/sycl/sycl/doc/EnvironmentVariables.md#oneapi_device_selector
+
+    // discardFilter = "!acceptFilter"
+    //  acceptFilter = "backend:filterStrings"
+    // filterStrings = "filterString[,filterString[,...]]"
+    //  filterString = "root[.sub[.subsub]]"
+    //          root = "*|int|cpu|gpu|fpga"
+    //           sub = "*|int"
+    //        subsub = "*|int"
+
+    // validation regex for filterString (not used in this code)
+    std::regex validation_pattern("^("
+            "\\*" // C++ escape for \, regex escape for literal '*'
+            "|"
+            "cpu" // ensure case-insenitive, when using
+            "|"
+            "gpu" // ensure case-insenitive, when using
+            "|"
+            "fpga" // ensure case-insenitive, when using
+            "|"
+            "[[:digit:]]+" // '<num>'
+            "|"
+            "[[:digit:]]+\\.[[:digit:]]+" // '<num>.<num>'
+            "|"
+            "[[:digit:]]+\\.\\*" // '<num>.*.*'
+            "|"
+            "\\*\\.\\*" // C++ and regex escapes, literal '*.*'
+            "|"
+            "[[:digit:]]+\\.[[:digit:]]+\\.[[:digit:]]+" // '<num>.<num>.<num>'
+            "|"
+            "[[:digit:]]+\\.[[:digit:]]+\\.\\*" // '<num>.<num>.*'
+            "|"
+            "[[:digit:]]+\\.\\*\\.\\*" // '<num>.*.*'
+            "|"
+            "\\*\\.\\*\\.\\*" // C++ and regex escapes, literal '*.*.*'
+        ")$", std::regex_constants::icase);
```

Comment:
We recently removed use of `std::regex` from the OpenCL adapter in #1039 due to profiling with [PIN](https://www.intel.com/content/www/us/en/developer/articles/tool/pin-a-dynamic-binary-instrumentation-tool.html) showed it to produce incredibly innefficient code. That was only for the OpenCL adapter, this would be on the critical path which makes me a bit uncomfortable.

Reply:
That (don't use `std::regex`) is good to know, thanks.

Currently, that regex pattern is only referenced in this comment:
line 491 `// TODO -- use regex validation_pattern to catch all other syntax errors in the ODS string`

Without this catch-all, some syntax errors will make it into the code that assumes there are no syntax errors, which is not great.

I would not expect that this code would ever be on the critical path of a performance critical loop. I would not expect users to query the list of devices (directly or indirectly) in a performance critical situation. Do you anticipate user code that does this?

To mitigate such user code (if it is a design goal), the parsing of the env var (including creating and using the regex) could be done only once and the outcome could be cached in the loader. The current code does not implement that, but it could.

Reply:
> Currently, that regex pattern is only referenced in this comment:
line 491 // TODO -- use regex validation_pattern to catch all other syntax errors in the ODS string

Is the regex only intended to be used during validation or as part of the parsing of the ODS string?

> To mitigate such user code (if it is a design goal), the parsing of the env var (including creating and using the regex) could be done only once and the outcome could be cached in the loader. The current code does not implement that, but it could.

I think this could be acceptable.